ghga_connector/core/file_operations.py

```diff
@@ -0,0 +1,44 @@
+# Copyright 2021 - 2022 Universität Tübingen, DKFZ and EMBL
+# for the German Human Genome-Phenome Archive (GHGA)
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+"""
+Contains Calls of the Presigned URLs in order to Up- and Download Files
+"""
+
+import pycurl
+
+from .exceptions import BadResponseCodeError, RequestFailedError
+
+
+def download_file(download_url, output_file):
+    """Download File"""
+
+    with open(output_file, "wb") as file:
+        curl = pycurl.Curl()
+        curl.setopt(curl.URL, download_url)
+        curl.setopt(curl.WRITEDATA, file)
+        try:
+            curl.perform()
+        except pycurl.error as pycurl_error:
+            raise RequestFailedError(download_url) from pycurl_error
+
+        status_code = curl.getinfo(pycurl.RESPONSE_CODE)
+        curl.close()
+
+    if status_code == 200:
+        return
```

Comment:
Theoretically all status codes in the 2xx range are success.
Is 200 the only valid thing that can be returned (currently)?

Reply:
Different success HTTP codes mean different things and usually are well defined. Therefore we must only accept the specific HTTP code we are expecting - in this case 200.

For example, when calling the download controller for a presigned download url, both 200 and 202 are 'success' answers, but only one of them returns the actual download url, while the other one tells us to retry later.

https://github.com/ghga-de/ghga-connector/blob/34eadc7365884504916baf33767700fe41d68e93/ghga_connector/core/api_calls.py#L110